choice observers, prep

nickchapman-da opened the pull request:
This PR contains preparatory work for _choice-observers_ #7709. The changes are simple, but wide ranging through the code base. They stop short of adding the new surface syntax in DAML. This will be done in a following PR.

These changes add support for _choice-observers_ to:

- Haskell AST -- optional `chcObservers` Expr
- Haskell `featureChoiceObservers`
- Haskell LF decode & encode
- Haskell type checker
- daml_lf_1.proto -- optional `observers` field in TemplateChoice message
- Scala AST  -- optional `choiceObservers` Expr
- Scala LF decode & encode
- Scala LF parser -- extend syntax for optional `observers` on `choice`s
- Scala LF type checker
- Scala speedy compiler -- pay attention to optional `choiceObservers`
- Scala speedy builtin `SBUBeginExercise` -- process choice-observers argument
- Scala engine -- construct transaction with choice-observers
- Scala transaction Node/NodeInfo
- Scala transaction version -- `minChoiceObservers`
- Scala transaction coder (encode & decode)
- transaction.proto -- new optional, repeated `observers` field

Things to be done in following PRs:

- Update specs: `daml-lf-1.rst` and `transaction.rst`
- Extend the surface syntax of DAML for choice-observers.
- Testing of which parties are informed. (Such tests will be easier using the surface syntax).

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

hurryabit commented on daml-lf/transaction/src/main/protobuf/com/daml/lf/transaction.proto:
```diff
@@ -94,6 +94,7 @@ message NodeExercise {
     com.daml.lf.value.VersionedValue return_value = 12;
     com.daml.lf.value.VersionedValue contract_key = 13; // optional
     KeyWithMaintainers key_with_maintainers = 14; // optional
+    repeated string observers = 15;
```
We should document that this need to be empty for whatever is the current version and below. We also need to agree if we want to call this `observers` or `informees`. cc @remyhaemmerle-da @bame-da @andreaslochbihler-da

hurryabit commented on daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:
```diff
@@ -227,6 +229,7 @@ object Node {
     *     actual validation stage.
     */
   final case class NodeExercises[+Nid, +Cid, +Val](
+      observers: Set[Party], //additional choice observers. -- NICK: re-order after signatories
```
When computing the witnesses of an exercise node, you'll need to consider the template observers and the choice observers. Thus, I would prefer to call this field `choiceObservers` (or `informees` if we go with that) to avoid any confusion.

@remyhaemmerle-da Do we need to make this field optional for the sake of backward compatibility?

hurryabit replied:
I would not speak of "additional choice observers". It should either be "choice observers" or "additional informees".

hurryabit commented on daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/NodeInfo.scala:
```diff
@@ -66,14 +66,15 @@ object NodeInfo {
+        stakeholders | actingParties | observers
       else
-        signatories | actingParties
+        signatories | actingParties | observers
```
Why are we adding the observers for a non-consuming choice? That's the sort of confusion I was referring to above.

remyhaemmerle-da commented on daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:
```diff
@@ -358,7 +358,9 @@ private[validation] object Typing {
             checkExpr(controllers, TParties)
           }
-          checkExpr(observers, TParties) // NICK, should this be conditional on supportsContractObservers?
+          observers.foreach {
+            checkExpr(_, TParties) // NICK, should this be conditional on supportsContractObservers?
+          }
```
```suggestion
          observers.foreach(checkExpr(_, TParties))
```
Don't need the comment anymore.

remyhaemmerle-da commented on daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala:
```diff
@@ -186,7 +186,10 @@ private[validation] object TypeTraversable {
           update) =>
         foreach(controllers, f)
-        foreach(observers, f)
+        observers match {
+          case Some(observers) => foreach(observers, f)
+          case None => ()
+        }
```
```suggestion
        observers.foreach(foreach(_, f))
```

remyhaemmerle-da commented on daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ModParser.scala:
```diff
@@ -141,6 +141,7 @@ private[parser] class ModParser[P](parameters: ParserParameters[P]) {
           name,
           !choiceTags(nonConsumingTag),
           controllers,
+          observers = None, //ENil(TBuiltin(BTParty)), //NICK, extend parser for (optional?) observers.
```
If you do not plan to fix the parser in this PR, please add a FIXME with a github issue.

remyhaemmerle-da commented on daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:
```diff
@@ -188,6 +188,7 @@ object TransactionBuilder {
       byKey: Boolean = true,
   ): Exercise =
     Exercise(
+      observers = Set.empty,
```
We probably have to add the `observers` to the argument of the definition. 
The question is if we should provide a default value. I think `Set.empty[String]` as a default value make sens.

hurryabit replied:
I'd prefer to not have a default value. I think they are just too dangerous.

remyhaemmerle-da replied:
Note that this is a library for writing test, and it may used by external code. 
That said, I am OK with no default value. In any case the `observers` should be an argument of the method.

remyhaemmerle-da commented on daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:
```diff
@@ -217,6 +217,7 @@ object TransactionCoder {
             .addAllChildren(ne.children.map(encodeNid.asString).toList.asJava)
             .addAllSignatories(ne.signatories.toSet[String].asJava)
             .addAllStakeholders(ne.stakeholders.toSet[String].asJava)
+            .addAllObservers(ne.observers.toSet[String].asJava)
```
We should check the version here ! 

You have to add in `TransactionVersions`: 
```scala
private[transaction] val minChoiceObserver = TransactionVersion("11")
```

And somewhere around here: 
```scala
if (transactionVersion precedes minChoiceObserver && ne.observers.nonEmpty) 
  throw  EncodeError(
                s"As of version $minNoControllers, the observer of an exercise node _must_ be empty",
  )
```

remyhaemmerle-da commented on daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:
```diff
@@ -420,11 +421,13 @@ object TransactionCoder {
           }
           signatories <- toPartySet(protoExe.getSignatoriesList)
           stakeholders <- toPartySet(protoExe.getStakeholdersList)
+          observers <- toPartySet(protoExe.getObserversList)
```
We should check the version !

remyhaemmerle-da commented on daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:
```diff
@@ -656,11 +659,13 @@ object TransactionCoder {
           actingParties_ <- toPartySet(protoExe.getActorsList)
           signatories_ <- toPartySet(protoExe.getSignatoriesList)
           stakeholders_ <- toPartySet(protoExe.getStakeholdersList)
+          observers_ <- toPartySet(protoExe.getObserversList)
```
We shoudl check the version here as well.

remyhaemmerle-da reviewed (commented):
Looks promising.

hurryabit commented on daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala:
```diff
@@ -122,8 +122,16 @@ object Util {
     choice match {
-      case TemplateChoice(name, consuming, _, selfBinder, argBinder, returnType, _) =>
-        TemplateChoiceSignature(name, consuming, (), selfBinder, argBinder, returnType, ())
+      case TemplateChoice(name, consuming, _, _, selfBinder, argBinder, returnType, _) =>
+        TemplateChoiceSignature(
+          name,
+          consuming,
+          (),
+          Some(()),
```
I'm in favor of preserving the optionality.

hurryabit commented on compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:
```diff
@@ -810,6 +810,8 @@ data TemplateChoice = TemplateChoice
     -- ^ The controllers of the choice. They have type @List Party@ and the
     -- template parameter in scope, and (since 1.2) also the choice parameter.
+  , chcObservers :: !(Maybe Expr)
+    -- ^ The controllers of the choice. They have type @List Party@
```
```suggestion
    -- ^ The observers of the choice. When they are present, they have type @List Party@.
```

hurryabit commented on compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs:
```diff
@@ -319,11 +319,17 @@ decodeChoice LF1.TemplateChoice{..} =
     <*> decodeName ChoiceName templateChoiceName
     <*> pure templateChoiceConsuming
     <*> mayDecode "templateChoiceControllers" templateChoiceControllers decodeExpr
+    <*> decodeExprMaybe templateChoiceObservers
```
```suggestion
    <*> traverse decodeExpr templateChoiceObservers
```
and drop the function below.

hurryabit commented on compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:
```diff
@@ -656,13 +656,19 @@ checkDefValue (DefValue _loc (_, typ) _noParties (IsTest isTest) expr) = do
   checkType paramType KStar
   checkType retType KStar
-  introExprVar param paramType $ checkExpr actors (TList TParty)
+  introExprVar param paramType $ checkExpr controllers (TList TParty)
+  introExprVar param paramType $ checkOptionalChoiceObservers observers
```
Rename `observers` above to `mbObservers` and
```suggestion
  introExprVar param paramType $ do
    checkExpr controllers (TList TParty)
    whenJust mbObservers $ \observers -> do
      -- check version is recent enough here
      checkExpr observers (TList TParty)
```
and drop `checkOptionalChoiceObservers` below.

hurryabit commented on compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:
```diff
@@ -655,6 +655,7 @@ convertChoice env tbinds (ChoiceData ty expr)
         , chcName = choiceName
         , chcConsuming = consuming == Consuming
         , chcControllers = controllers `ETmApp` EVar this `ETmApp` EVar arg
+        , chcObservers = Nothing -- NICK, need syntax for non-empty choice-observers
```
Let's do that in a separate PR.

hurryabit commented on daml-lf/language/src/test/scala/com/digitalasset/daml/lf/language/AstSpec.scala:
```diff
@@ -125,6 +125,7 @@ class AstSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
       name = name,
       consuming = true,
       controllers = eParties,
+      observers = None,
```
We will probably need a new test for the `Some` case here as well.

hurryabit commented on daml-lf/parser/src/test/scala/com/digitalasset/daml/lf/testing/parser/ParsersSpec.scala:
```diff
@@ -475,6 +475,7 @@ class ParsersSpec extends WordSpec with TableDrivenPropertyChecks with Matchers
               name = n"Sleep",
               consuming = true,
               controllers = e"Cons @Party [person] (Nil @Party)",
+              observers = None, // NICK, was: e"Nil @Party",
```
We'll need a test for the `Some` case here as well.

hurryabit commented on ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoDivulgenceSpec.scala:
```diff
@@ -56,6 +56,7 @@ private[dao] trait JdbcLedgerDaoDivulgenceSpec extends LoneElement with Inside {
       val builder = TransactionBuilder()
       val rootExercise = builder.add(
         NodeExercises(
+          observers = Set.empty,
```
Given that this test is about divulgence and choice observers are too, we'll need some new tests here.

hurryabit reviewed (commented):
I've left a mixture of wining about missing tests and Haskell pro tips. 😃

nickchapman-da commented:
@Remy @Martin I've worked though all your comments (including the 2nd set from Martin). I tried not to miss anything. When I haven't made the change already, I added a NICK/FIXME marker in the code. I updated the desc to say what remains to be done before the first PR can be extracted.

One question:
- Should choice observers be added to the informees for a non-consuming choice?

hurryabit commented on daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:
```diff
@@ -922,10 +923,11 @@ private[lf] object SBuiltin {
       }
       val sigs = extractParties(args.get(3))
-      val obs = extractParties(args.get(4))
+      val contractObs = extractParties(args.get(4))
```
Given that we need to change it anyway, let's spell it out. I also think speaking of "contract observers" and "choice observers" is somewhat inconsistent. It should either be "template observers" and "choice observers" or "contract observers" and "exercise observers". I heavily prefer the former.
```suggestion
      val templateObservers = extractParties(args.get(4))
```

andreaslochbihler-da commented:
>     * Should choice observers be added to the informees for a non-consuming choice?

Yes. @oggy- will update the ledger model accordingly once all this has landed.

remy commented:
Wrong "@remy". I'll unsubscribe myself…

nickchapman-da commented:
> > ```
> > * Should choice observers be added to the informees for a non-consuming choice?
> > ```
> 
> Yes. @oggy- will update the ledger model accordingly once all this has landed.

Sorry. I didn't ask the question I intended!
Of course _choice observers_ must be _informees_ for a *non-consuming* choice. That's their purpose.
I meant to ask: should they be _informees_ of a *consuming* choice? I think the answer is yes.

nickchapman-da commented:
> Wrong "@remy". I'll unsubscribe myself…

Sorry. I meant @remyhaemmerle-da

andreaslochbihler-da commented:
> I meant to ask: should they be _informees_ of a _consuming_ choice? I think the answer is yes.

Yes. IIRC choice observers can be specified for each choice individually. So if the choice observers were not added to the informees of a consuming exercise, then specifying them on a consuming choice (which is determined statically in DAML) would be pointless.

hurryabit commented on compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:
```diff
@@ -127,6 +127,13 @@ featureToTextContractId = Feature
+featureChoiceObservers :: Feature  -- issue #7709
+featureChoiceObservers = Feature
+    { featureName = "Choice observers"
+    , featureMinVersion = versionDev
```
Can you please leave a comment
```
- TODO Change as part of #7139
```
(like above) and add choice observers to that issue.

nickchapman-da replied:
done

nickchapman-da commented on daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala:
```diff
@@ -39,6 +39,7 @@ private[parser] object Lexer extends RegexParsers {
     "fetch_by_key" -> `fetch_by_key`,
     "lookup_by_key" -> `lookup_by_key`,
     "by" -> `by`,
+    "ob" -> `ob`,
```
For templates, "observers" is lexed as an identifier, not a keyword, and so requires additional `,`s to separate the various fields:
```
      (Id("precondition") ~> expr) ~
      (`,` ~> Id("signatories") ~> expr) ~
      (`,` ~> Id("observers") ~> expr) ~
      (`,` ~> Id("agreement") ~> expr) ~
```

whereas for choices we have this keyword style: `by`, `to`, which doesn't require commas.
(this is why I went for a new `ob` keyword)

```
      `by` ~ expr ~
      opt(`ob` ~> expr) ~
      `to` ~ expr ^^ {
```

If you want to move to the identifier style, then we should probably also change keyword `by` to identifier `controllers` to be consistent, and so require commas between all the fields.

nickchapman-da replied:
I pushed a change to use identifier "observers" syntax (and leaving the `by` keyword as it is)
so you can see if you are happy with what it looks like
or decide how you want it to change
@hurryabit @remyhaemmerle-da

hurryabit replied:
I'd be fine with moving from `by` to `controllers` and adding some commas. Could we do that in a prep PR?

nickchapman-da replied:
done

hurryabit reviewed (commented):
I'm happy with the changes on the Haskell side and the API changes on the Scala side. @remyhaemmerle-da can you please approve this PR once you're happy with the rest.

remyhaemmerle-da commented on daml-lf/spec/daml-lf-1.rst:
```diff
@@ -4163,6 +4163,12 @@ The type checker will reject any DAML-LF < 1.2 program that tries to access
+Choice observers
+................
+
+  FIXME: https://github.com/digital-asset/daml/issues/7709
```
:+1:

remyhaemmerle-da reviewed (approved):
Looks good to me.

remyhaemmerle-da reviewed (approved):
I am fine with merging the PR such as.
